File: 02.11/baekjoon_22352_박동현.py
import sys; input = sys.stdin.readline


def dfs(x,y,now,nxt):
    if before[x][y] != now: return
    before[x][y] = nxt
    for dx,dy in dr:
        nx,ny = x+dx, y+dy
        if 0 <=nx < N and 0 <= ny < M:
            dfs(nx,ny,now,nxt)

def check():
    cnt = 0
    for i in range(N):
        for j in range(M):
            if before[i][j] != after[i][j]:
                dfs(i,j, before[i][j], after[i][j])
                cnt += 1
                if cnt > 1: return False
    return True

dr = (1,0),(0,1),(-1,0),(0,-1)

N,M = map(int,input().split())
before = [[*map(int,input().split())] for _ in range(N)]
after = [[*map(int,input().split())] for _ in range(N)]

print("YES" if check() and before==after else "NO")

File: 02.11/baekjoon_22352_박동현.go
package main

import (
	"bufio"
	"os"
	"slices"
	"strconv"
)

var (
	dx = [4]int{0, 1, 0, -1}
	dy = [4]int{1, 0, -1, 0}
)

func fillArr(arr *[][]int, size int) {
	for i := 0; i < len(*arr); i++ {
		(*arr)[i] = make([]int, size)
		for j := 0; j < size; j++ {
			(*arr)[i][j] = input()
		}
	}
}

func isEqual(arr, brr [][]int) bool {
	for i := 0; i < len(arr); i++ {
		if !slices.Equal(arr[i], brr[i]) {
			return false
		}
	}
	return true
}

func dfs(N, M, x, y, now, nxt int, arr *[][]int) {
	if (*arr)[x][y] != now {
		return
	}
	(*arr)[x][y] = nxt

	for i := 0; i < 4; i++ {
		nx, ny := x+dx[i], y+dy[i]
		if 0 <= nx && nx < N && 0 <= ny && ny < M {
			dfs(N, M, nx, ny, now, nxt, arr)
		}
	}
}

func main() {
	bs.Split(bufio.ScanWords)
	defer bw.Flush()

	N, M := input(), input()
	before := make([][]int, N)
	fillArr(&before, M)
	after := make([][]int, N)
	fillArr(&after, M)
	cnt := 0
	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			if before[i][j] != after[i][j] {
				dfs(N, M, i, j, before[i][j], after[i][j], &before)
				cnt++
				if cnt > 1 {
					bw.WriteString("NO")
					return
				}
			}
		}
	}
	if isEqual(before, after) {
		bw.WriteString("YES")
	} else {
		bw.WriteString("NO")
	}
}

// fastIO
var (
	bs = bufio.NewScanner(os.Stdin)
	bw = bufio.NewWriter(os.Stdout)
)

func input() int {
	bs.Scan()
	num, _ := strconv.Atoi(bs.Text())
	return num
}
